src/query/src/dist_plan/analyzer/utils.rs

```diff
@@ -12,188 +12,514 @@
+fn get_alias_layer_from_node(node: &LogicalPlan) -> DfResult<AliasLayer> {
+    match node {
+        LogicalPlan::Projection(proj) => Ok(get_alias_layer_from_exprs(&proj.expr)),
+        LogicalPlan::Aggregate(aggr) => Ok(get_alias_layer_from_exprs(&aggr.group_expr)),
+        LogicalPlan::SubqueryAlias(subquery_alias) => {
+            let mut layer = AliasLayer::default();
+            let old_columns = subquery_alias.input.schema().columns();
+            for old_column in old_columns {
+                let new_column = Column::new(
+                    Some(subquery_alias.alias.clone()),
+                    old_column.name().to_string(),
+                );
+                // mapping from old_column to new_column
+                layer.insert_alias(old_column, HashSet::from([new_column]));
+            }
+            Ok(layer)
+        }
+        LogicalPlan::TableScan(scan) => {
+            let columns = scan.projected_schema.columns();
+            let mut layer = AliasLayer::default();
+            for col in columns {
+                layer.insert_alias(col.clone(), HashSet::from([col.clone()]));
+            }
+            Ok(layer)
+        }
+        _ => {
+            let input_schema = node
+                .inputs()
+                .first()
+                .ok_or_else(|| {
+                    datafusion::error::DataFusionError::Internal(
+                        "only accept plan with at most one child".to_string(),
+                    )
+                })?
+                .schema();
+            let output_schema = node.schema();
+            // only accept at most one child plan, and if not one of the above nodes,
+            // also shouldn't modify the schema or else alias scope tracker can't support them
+            if node.inputs().len() > 1 {
+                Err(datafusion::error::DataFusionError::Internal(
+                    "only accept plan with at most one child".to_string(),
+                ))
+            } else if node.inputs().len() == 1 {
+                if input_schema != output_schema {
+                    let input_columns = input_schema.columns();
+                    let all_input_is_in_output = input_columns
+                        .iter()
+                        .all(|c| output_schema.is_column_from_schema(c));
+                    if all_input_is_in_output {
+                        // all input is in output, so it's just adding some columns, we can do identity mapping for input columns
+                        let mut layer = AliasLayer::default();
+                        for col in input_columns {
+                            layer.insert_alias(col.clone(), HashSet::from([col.clone()]));
+                        }
+                        Ok(layer)
+                    } else {
+                        // otherwise use the intersection of input and output
+                        // TODO(discord9): maybe just make this case unsupported for now?
```

Comment:
This part about tracking in node with different input/ouput schema that's not table scan/projection/subquery alias might be somewhere handled better but I don't know how